synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -466,6 +466,27 @@ def _flatten_dict(
     prefix: Optional[List[str]] = None,
     result: Optional[Dict[str, str]] = None,
 ) -> Dict[str, str]:
+    """
+    Given a JSON dictionary (or event) which might contain sub dictionaries,
+    flatten it into a single layer dictionary by combining the keys & sub-keys.
+
+    Transforms:
+
+        {"foo": {"bar": "test"}}
+
+    To:
+
+        {"foo.bar": "test"}
```

Comment:
How would (should) the dict 

```python
{"foo.bar": 1, "foo": {"bar": 2}}
```

be transformed? I imagine that `2` will win... but maybe we should just say something like "the result is undefined if there is any kind of conflict between keys"?

Reply:
This is undefined behavior, one of which the pull requests that I'll be implementing will be fixing.

Reply:
MSC3873 for the record.